Node names in bootstrap config have no ports

DaveCTurner opened the pull request:
In cases where node names and transport addresses can be muddled, it is unclear
that `cluster.initial_master_nodes: master-a:9300` means to look for a node
called `master-a:9300` rather than a node called `master-a` with transport port
`9300`. This commit adds docs to that effect.

DaveCTurner committed:
Node names in bootstrap config have no ports

In cases where node names and transport addresses can be muddled, it is unclear
that `cluster.initial_master_nodes: master-a:9300` means to look for a node
called `master-a:9300` rather than a node called `master-a` with transport port
`9300`. This commit adds docs to that effect.

elasticmachine commented:
Pinging @elastic/es-distributed

andrershov reviewed (commented):
Do we want to forbid the usage ":" in node names not to create confusion?

DaveCTurner commented:
It's tricky, but I don't think we should do so.

I can see a certain amount of value in being able to name your nodes `host-a:9300` and `host-a:9301`, even though this could be deeply confusing. I had to try quite hard to avoid wording that implied that you weren't allowed to do that.

ywelsch commented on docs/reference/modules/discovery/bootstrapping.asciidoc:
```diff
@@ -47,9 +47,11 @@ cluster.initial_master_nodes:
+there is more than one {es} node with the same IP address then the IP addresses
+can be distinguished by appending each node's transport port as shown below.
+However, {es} nodes should have unique node names so {es} does not allow two
+nodes with the same name to be distinguished by their transport port like this.
```
I find this last sentence here very confusing, with what's presented as a logical conclusion (... so {es} does not allow ...). The problem we're trying to address is that users  think  that this setting is resolved in the usual way (i.e. same as `discovery.seed_hosts`), but we require exact matching on this one, in particular `host-name:port` is not an acceptable combination.

I wonder if we should just say that the values of this setting are not resolved and need to exactly match either the node's name (which defaults to the host name), or the publish address, which can be presented as an IP address or an IP:PORT pair. We can even add a sentence to the effect that it is generally preferable to identify nodes uniquely through their node names.

DaveCTurner replied:
Ok, I had clearly spent too long staring at this, and now I don't like this wording either. This setting is a pain to document. I have tried again in 61b5f6c.

DaveCTurner committed:
Merge branch 'master' into 2019-04-26-no-transport-ports-with-node-names

DaveCTurner committed:
Rework

ywelsch commented on docs/reference/modules/discovery/bootstrapping.asciidoc:
```diff
@@ -10,11 +10,17 @@ data folder and freshly-started nodes that are joining an existing cluster
+- The IP address of the node's <<modules-transport,publish address>> which is
+  normally the IP address to which <<common-network-settings,`network.host`>>
+  resolves but <<advanced-network-settings,this can be overridden>>.
+- The IP address and port of the node's publish address, in the form `IP:PORT`.
```
sorry, I still don't like this. Instead I suggest something along the lines of:

```
- The <<node.name,node name>> of the node, which defaults to the node's hostname if
  `node.name` is not set.
- The IP address of the node's <<modules-transport,publish address>>, which is the IP
  address that this node announces itself to other nodes and which the other nodes use to
  communicate with this node. In case where multiple nodes are running on the same host and
  using the same publish address, this can be further qualified with the port in the form
  `IP:PORT`. Setting unique node names and configuring cluster bootstrapping via the
  node names should however be preferred.
```

DaveCTurner replied:
Are you sure we want to express this kind of preference for `node.name` over the IP address? Why do we have this preference? It seems like that might push some users down a path of trying harder than necessary to pin down their node names simply to avoid the non-preferred, but easier-for-them, IP addresses.

ywelsch replied:
Selecting proper node names should be a good practice (even it is just consciously not setting it and leaving it to the default, i.e., the host name). Given this great identifier of nodes, why not  use it instead of something more cryptic like IPs?

That being said, my intention was for the last sentence ("Setting unique node names and configuring cluster bootstrapping via the node names should however be preferred") to apply for the cases where you run multiple nodes on the same host ("In case where multiple nodes are running on the same host and using the same publish address..."). In that case, it's truly important to use unique node names to avoid confusion.

DaveCTurner replied:
Ok, I added qualifiers of the form `if it is not possible to use the `node.name` of the node`.

I would like to keep the four options as four separate things, rather than combining them down to two as you suggested. My thinking is that users should generally pick one of these four strategies consistently across their cluster; if we only give two options with embedded conditionals then I think it's harder to see the full choice that needs to be made.

ywelsch commented on docs/reference/setup/important-settings/discovery-settings.asciidoc:
```diff
@@ -68,7 +69,8 @@ cluster.initial_master_nodes:
 <5> If multiple master nodes share an IP address then the transport port must
-    be used to distinguish between them.
+    be used to distinguish between them. Nodes should have unique names so you
+    cannot distinguish named nodes by port.
```
I think we should just scratch the IP:PORT in this "getting started" section.

DaveCTurner replied:
Good idea, yes.

ywelsch commented on docs/reference/modules/discovery/bootstrapping.asciidoc:
```diff
@@ -47,17 +53,16 @@ cluster.initial_master_nodes:
-You can use a mix of IP addresses and node names too. If there is more than one
-Elasticsearch node with the same IP address then the transport port must also
-be given to specify exactly which node is meant:
+You can also use IP addresses, or IP addresses and ports, or even a mix of IP
```
While node names are generally preferred, you can also use IP addresses, or IP addresses and ports, or even a mix of IP addresses and node names:

DaveCTurner replied:
Ok, qualified this with `If it is not possible to use the names of the nodes...`.

DaveCTurner committed:
Merge branch 'master' into 2019-04-26-no-transport-ports-with-node-names

DaveCTurner committed:
Indicate preference for node name over IP address

DaveCTurner committed:
Remove mention of IP addresses from getting-started

ywelsch commented on docs/reference/modules/discovery/bootstrapping.asciidoc:
```diff
@@ -10,11 +10,20 @@ data folder and freshly-started nodes that are joining an existing cluster
+node:
+
+- The <<node.name,node name>> of the node.
+- The node's hostname if `node.name` is not set, because `node.name` defaults
```
Having this as a separate item will I think only lead to confusion (which you already go at  great length of clearing up in `discovery-settings.asciidoc`) that this thing will somehow be resolved similar to `seed_hosts`.

ywelsch reviewed (approved):
I've left an optional comment.

jaymode committed:
Update TLS ciphers and protocols for JDK 11 (#41808)

* Update TLS ciphers and protocols for JDK 11 (#41385)

This commit updates the default ciphers and TLS protocols that are used
after the minimum supported JDK is JDK 11. The conditionals around
TLSv1.3 and 256-bit cipher support have been removed. JDK 11 no longer
requires an unlimited JCE policy file for 256 bit cipher support and
TLSv1.3 is supported in JDK 11+. New cipher support has been introduced
in the newer JDK versions as well. The ciphers are ordered with PFS
ciphers being most preferred, then AEAD ciphers, and finally those with
mainstream hardware support.

* Fixes for TLSv1.3 on JDK11

* fix for JDK-8212885

romseygeek committed:
Correct spelling of MockLogAppender.PatternSeenEventExpectation (#41893)

The class was called PatternSeenEventExcpectation. This commit
is a straight class rename to correct the spelling.

DaveCTurner committed:
Handle serialization exceptions during publication (#41781)

Today if an exception is thrown when serializing a cluster state during
publication then the master enters a poisoned state where it cannot publish any
more cluster states, but nor does it stand down as master, yielding repeated
exceptions of the following form:

```
failed to commit cluster state version [12345]
org.elasticsearch.cluster.coordination.FailedToCommitClusterStateException: publishing failed
        at org.elasticsearch.cluster.coordination.Coordinator.publish(Coordinator.java:1045) ~[elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.cluster.service.MasterService.publish(MasterService.java:252) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:238) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:142) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:681) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-7.0.0.jar:7.0.0]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_144]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_144]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
Caused by: org.elasticsearch.cluster.coordination.CoordinationStateRejectedException: cannot start publishing next value before accepting previous one
        at org.elasticsearch.cluster.coordination.CoordinationState.handleClientValue(CoordinationState.java:280) ~[elasticsearch-7.0.0.jar:7.0.0]
        at org.elasticsearch.cluster.coordination.Coordinator.publish(Coordinator.java:1030) ~[elasticsearch-7.0.0.jar:7.0.0]
        ... 11 more
```

This is because it already created the publication request using
`CoordinationState#handleClientValue()` but then it fails before accepting it.
This commit addresses this by performing the serialization before calling
`handleClientValue()`.

Relates #41090, which was the source of such a serialization exception.

lcawl committed:
[DOCS] Fix callouts for dataframe APIs (#41904)

DaveCTurner committed:
Merge branch 'master' into 2019-04-26-no-transport-ports-with-node-names

DaveCTurner committed:
Link to FQDNs note

DaveCTurner committed:
WIP fix up force-push

DaveCTurner committed:
Merge branch 'master' into 2019-04-26-no-transport-ports-with-node-names
